library(GenomicFeatures)
library(dplyr)
library(pheatmap)
library(RColorBrewer)
library(ggplot2)
library(tibble)
library(edgeR)
library(readxl)
library(DESeq2)


rm(list = ls())

# Fabio - FCD type IIa and IIb passing quality control
# Generated by PCA_analysis_FCD_Fabio.R
load(file = "FCD_Fabio.rds")

pheno_br <- pheno %>% dplyr::select("sample", "Subtype") %>%
  dplyr::mutate("Cohort" = "Br")


# Aus - FCD type IIa and IIb passing quality control
pheno_aus <- read.table("./sample_annotation/samplesheet_FCD_Australia.csv", stringsAsFactors = FALSE,
                    header = TRUE, sep = ",")
pheno_aus <- pheno_aus %>% dplyr::select(c("sample", "Diagnosis", "disease")) %>% distinct() %>%
  mutate(Subtype = case_when(
    Diagnosis == 'IIb' ~ 'FCD IIb',
    Diagnosis == 'IIa' ~ 'FCD IIa',
    Diagnosis == 'Ia X Y N' ~ 'FCD Ia',
    Diagnosis == 'Non-epilepsy' ~ 'Control',
    Diagnosis == 'HS - Epilepsy' ~ 'TLE+HS')) %>%
  dplyr::select("sample", "Subtype") %>%
  dplyr::mutate("Cohort" = "Aus") %>%
  dplyr::filter(Subtype %in% c("FCD IIa", "FCD IIb"))
head(pheno_aus)

rsem_aus <- read.table("./RNA-seq_FCD_Aus/rsem.merged.gene_counts.tsv", stringsAsFactors = FALSE, header = TRUE)
rownames(rsem_aus) <- rsem_aus$gene_id
rsem_aus <- rsem_aus[, c(-1, -2)]
rsem_aus <- rsem_aus[, colnames(rsem_aus) %in% pheno_aus$sample]
head(rsem_aus)

rsem_comb <- cbind(rsem_aus, rsem_counts)
head(rsem_comb)

pheno_comb <- rbind(pheno_aus, pheno_br)
rownames(pheno_comb) <- pheno_comb$sample

sample_id <- pheno_comb$sample
NS <- length(sample_id)
#order according to sample_annot
perm <- c()
for (i in 1:length(sample_id)) {
  perm <- c(perm, grep(sample_id[i], colnames(rsem_comb)))
}
rsem_comb <- rsem_comb[,perm]


# DESeq2 log2 normalization (vsd) and batch correction
dds <- DESeqDataSetFromMatrix(countData = round(rsem_comb),
                              colData = pheno_comb %>% dplyr::select(c("Cohort", "Subtype")),
                              design= ~ Cohort + Subtype)
keep <- rowSums(counts(dds)) >= 5
dds <- dds[keep,]

range(counts(dds))
dim(counts(dds))
colData(dds)
levels(dds$Subtype)
levels(dds$Cohort)

# Remove batch effect
vsd <- vst(dds, blind=FALSE)
#rld <- rlog(dds, blind=FALSE)
head(assay(vsd), 3)
mat <- assay(vsd)
mm <- model.matrix(~Subtype, colData(vsd))
mat <- limma::removeBatchEffect(mat, batch=vsd$Cohort, design=mm)
assay(vsd) <- mat

head(mat)

# We estimate the variance for each row in the logcounts matrix
# var_genes <- apply(assay(vsd), 1, var)
# head(var_genes)

# Get the gene names for the top 1000 most variable genes
# select_var <- names(sort(var_genes, decreasing=TRUE))[1:5000]
# head(select_var)



# Sample distance heatmap
sampleDists <- dist( t( assay(vsd) ) )
# sampleDists <- dist( t( mat[rownames(mat) %in% select_var, ] ) )
# sampleDistMatrix <- as.matrix( sampleDists )
#colnames(sampleDistMatrix) <- NULL
colors <- colorRampPalette( rev(brewer.pal(9, "Blues")) )(255)

annotation_rows = data.frame(
  Subtype = factor(pheno_comb$Subtype),
  Cohort = factor(pheno_comb$Cohort)
)
rownames(annotation_rows) <- row.names(sampleDistMatrix)

ann_colors = list(
  #Donor = c("SLE392" = "tan", "SLE413" = "tan1", "SLE298" = "tan3", "SLE456" = "tan4"),
  Subtype = c("FCD IIa" = "purple", "FCD IIb" = "orange"),
  Cohort = c(Aus = "mistyrose1", Br = "mistyrose3")
)

pheatmap(sampleDistMatrix,
         clustering_distance_rows=sampleDists,
         clustering_distance_cols=sampleDists,
         annotation_row = annotation_rows,
         annotation_col = annotation_rows,
         show_colnames = F,
         show_rownames = F,
         annotation_colors = ann_colors,
         col=colors,
         clustering_method = "mcquitty",
         treeheight_row = 30,
         treeheight_col = 30
         
)

# Expression matrix normalized and corrected batch effect
expr_mat <- assay(vsd)
dim(expr_mat)
dim(pheno_comb)

save(expr_mat, rsem_comb, rsem_counts, pheno_comb, file = "FCD_Expression_BatchCorrection.rds")
